### PR TITLE
New addon Fist heroes win

### PR DESCRIPTION
**Pull request by frankystone**
This should be ready now.

Anyway i have trouble to test this in multiplayer lan games:

If i try to cheat and give other players a hero, the other players get no inbox message and by looking at the soldier stats for the other players, they also do not have a hero. Testing scenario is:

* Start two instances of widelands
* create a LAN game in one window
* in the other window join this lan game
* start game
* try to cheat and give another player a hero

In single player games all works as expected. Does cheating prevent something?

### CODE REVIEW

**Comment by Noordfrees** [comment]
Input on the cheating console is not sent to the other clients, so it changes the game state only on the client on which you're typing the cheat code. Effectively, a deliberately introduced desync. To get all the clients back in sync, save the game in the cheating client and start a new MP match from this savegame.

File: addons/first_heroes_win.wad/init.lua [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+local MAXHEROES = 4
+
+push_textdomain("first_heroes_win.wad", true)
+
+local wc_name = "First Heroes Win"
+-- This needs to be exactly like wc_name, but localized, because wc_name
+-- will be used as the key to fetch the translation in C++
+local wc_descname = _("First Heroes Win")
+local wc_version = 1
+local wc_desc = _("The player who has at least %1% fully trained soldiers in his warehouses wins the game."):bformat(MAXHEROES)
```
Comment (Noordfrees) [inline addons/first_heroes_win.wad/init.lua:32]: This should use `ngettext` or you'll be chased by unhappy translators ;)

File: addons/first_heroes_win.wad/init.lua [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+         opponent = opponent or nil
+         local title = ""
+         local msg = ""
+         if opponent then
+            title = _("Hero created!")
+            msg = msg .. p(_("%1% has created a hero!"):bformat(opponent.name))
+         else
+            title = _("You have a new Hero!")
+         end
+         msg = msg .. h2(_("Hero Statistic: "))
```
Comment (Noordfrees) [inline addons/first_heroes_win.wad/init.lua:97]: trailing whitespace

File: addons/first_heroes_win.wad/init.lua [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+-- This needs to be exactly like wc_name, but localized, because wc_name
+-- will be used as the key to fetch the translation in C++
+local wc_descname = _("First Heroes Win")
+local wc_version = 1
+local wc_desc = _("The player who has at least %1% fully trained soldiers in his warehouses wins the game."):bformat(MAXHEROES)
+local wc_first_hero = "Number of heroes"
+-- This needs to be exactly like wc_first_hero, but localized, because wc_first_hero
+-- will be used as the key to fetch the translation in C++
+local wc_first_hero_i18n = _("Number of heroes")
+
```
Comment (Noordfrees) [inline addons/first_heroes_win.wad/init.lua:37]:
```suggestion
   textdomain = "first_heroes_win.wad",

```

File: addons/first_heroes_win.wad/init.lua [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      -- Install statistics hook
+      hooks.custom_statistic = {
+         name = wc_first_hero,
+         pic = "images/wui/buildings/prefer_heroes.png",
+         calculator = function(p)
+            return plr_data[p.number].nr_heroes
+         end,
+      }
+
+      push_textdomain("first_heroes_win.wad", true)
```
Comment (Noordfrees) [inline addons/first_heroes_win.wad/init.lua:85]: This needs to be *inside* the body of the functions `_get_statistic_msg` and `_get_result_msg` in which the translatable strings are located

**Comment by frankystone** [comment]
Should all be fixed now.

Unsure about the ngettext() thing though.

File: addons/first_heroes_win.wad/init.lua [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+local textdomain = "first_heroes_win.wad"
+
+push_textdomain(textdomain, true)
+
+local wc_name = "First Heroes Win"
+-- This needs to be exactly like wc_name, but localized, because wc_name
+-- will be used as the key to fetch the translation in C++
+local wc_descname = _("First Heroes Win")
+local wc_version = 1
+local wc_desc = _("The player who has at least %1% fully trained soldiers in his warehouses wins the game."):bformat(ngettext("%1%", "%1%", MAXHEROES))
```
Comment (Noordfrees) [inline addons/first_heroes_win.wad/init.lua:34]: Not quite
```suggestion
local wc_desc = ngettext(
   "The player who has at least %1% fully trained soldier in his warehouses wins the game.",
   "The player who has at least %1% fully trained soldiers in his warehouses wins the game.",
   MAXHEROES):bformat(MAXHEROES)
```

File: addons/first_heroes_win.wad/init.lua [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+local wc_first_hero = "Number of heroes"
+-- This needs to be exactly like wc_first_hero, but localized, because wc_first_hero
+-- will be used as the key to fetch the translation in C++
+local wc_first_hero_i18n = _("Number of heroes")
+
+local r = {
+   name = wc_name,
+   description = wc_desc,
+   peaceful_mode_allowed = true,
```
Comment (Noordfrees) [inline addons/first_heroes_win.wad/init.lua:43]:
```suggestion
   peaceful_mode_allowed = true,
   textdomain = textdomain,
```

**Review by Noordfrees — COMMENTED** [review]
We're getting there ;)

File: addons/first_heroes_win.wad/init.lua (hunk 6 above)
Comment (frankystone) [reply addons/first_heroes_win.wad/init.lua:46]: Can you explain please?

Comment (Noordfrees) [reply addons/first_heroes_win.wad/init.lua:46]: When you load a savegame, the localized win condition name to be shown in the UI is fetched from the textdomain which is specified by this table. If this key is absent, the official `win_conditions` textdomain is used; add-ons need to specify their own textdomain here so the strings are still localized in the Load Game screen.

Comment (Noordfrees) [reply addons/first_heroes_win.wad/init.lua:46]: https://www.widelands.org/documentation/add-ons/#win-condition

**Comment by frankystone** [comment]
Thanks :+1:

**Comment by Noordfrees** [comment]
LGTM
Please re-run `UpdateList`, then this can be merged

**Pull request merged**